### PR TITLE
Add Parser for /proc/net/route

### PR DESCRIPTION
**Pull request by lyshar**
Working on a parser for IPv6 as well (it's a different file called `/proc/net/ipv6_route`)

Use pfs to get network device information from `/proc/net/route`. Adds parser, unit test, and sample out for new parser.

### CODE REVIEW

File: include/pfs/types.hpp [hunk 1]
```diff
@@ -604,6 +604,22 @@ struct id_map
     size_t length = UINT32_MAX;
 };
 
+// Based on https://elixir.bootlin.com/linux/v5.19.10/source/net/ipv4/fib_trie.c#L2998
```
Comment (dtrugman) [inline include/pfs/types.hpp:607]: Since line numbers and implementations change, and you don't always want to see a specific linux kernel version, maybe it's better to be consistent with other hints in the file and say:
`// Hint: See 'fib_route_seq_show @ fib_trie.c'`
This exact same function existed since 2.6.32 (https://elixir.bootlin.com/linux/v2.6.32/source/net/ipv4/fib_trie.c#L2567)

File: include/pfs/types.hpp [hunk 2]
```diff
@@ -604,6 +604,22 @@ struct id_map
...
+struct route
+{
+    std::string iface;
+    ip destination;
+    ip gateway;
+    unsigned int flags;
+    unsigned short refcnt;
```
Comment (dtrugman) [inline include/pfs/types.hpp:614]: Why `unsigned short`?
The value is always `0` according to the code, but the `printf` uses `%d`...

File: include/pfs/types.hpp [hunk 3]
```diff
@@ -604,6 +604,22 @@ struct id_map
     size_t length = UINT32_MAX;
 };
 
+// Based on https://elixir.bootlin.com/linux/v5.19.10/source/net/ipv4/fib_trie.c#L2998
+struct route
+{
+    std::string iface;
+    ip destination;
+    ip gateway;
+    unsigned int flags;
```
Comment (dtrugman) [inline include/pfs/types.hpp:613]: Have you considered going the extra mile and parsing the flags? You can find them [here](https://linux.die.net/man/8/route) and it should not be that hard to parse...

File: include/pfs/types.hpp [hunk 4]
```diff
@@ -604,6 +604,22 @@ struct id_map
...
+    unsigned short refcnt;
+    unsigned short use;
```
Comment (dtrugman) [inline include/pfs/types.hpp:615]: Why `unsigned short`?
The value is always `0` according to the code, but the `printf` uses `%u`...

File: include/pfs/types.hpp [hunk 5]
```diff
@@ -604,6 +604,22 @@ struct id_map
...
+    uint32_t metric;
```
Comment (dtrugman) [inline include/pfs/types.hpp:616]: I think that even though the struct has `uint32_t fib_priority`, because the `printf` uses `%d`, it's going to be more consistent to define that as `int`. wdyt?

File: include/pfs/types.hpp [hunk 6]
```diff
@@ -604,6 +604,22 @@ struct id_map
...
+    ip mask;
+    uint32_t mtu;
```
Comment (dtrugman) [inline include/pfs/types.hpp:618]: Same comment for `mtu`, `window` and `irtt` as above. We should be consistent with the `printf` formatter? Being consistent with the type might create discrepancies, being consistent with the formatter cannot.

File: include/pfs/procfs.hpp [hunk 7]
```diff
@@ -73,6 +73,8 @@ class procfs final
 
     std::string get_version_signature() const;
 
+    std::vector<route> get_routes() const;
```
Comment (dtrugman) [inline include/pfs/procfs.hpp:76]: Since the file is located under `/proc/net/route`, this should move under `net.hpp` and be called `get_route`

File: src/parsers/route.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+#include "pfs/types.hpp"
+
+#include <set>
+#include <sstream>
+
+namespace pfs {
+namespace impl {
+namespace parsers {
+
+route parse_route_line(const std::string& line)
```
Comment (dtrugman) [inline src/parsers/route.cpp:28]: Since the file is located under `/proc/net`, the file should be called `net_route.cpp` and the method `parse_net_route_line`

File: include/pfs/types.hpp [hunk 9]
```diff
@@ -604,6 +604,22 @@ struct id_map
     size_t length = UINT32_MAX;
 };
 
+// Based on https://elixir.bootlin.com/linux/v5.19.10/source/net/ipv4/fib_trie.c#L2998
+struct route
```
Comment (dtrugman) [inline include/pfs/types.hpp:608]: We should probably use `net_route` for consistency

File: src/parsers/route.cpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+{
+    // Some examples:
+    // clang-format off
+    // Iface   Destination     Gateway         Flags   RefCnt  Use     Metric  Mask            MTU     Window  IRTT
+    // eth0    00000000        07030301        0003    0       0       0       00000000        0       0       0
+    // eth0    07030301        00000000        0001    0       0       0       00F0FFFF        0       0       0
```
Comment (dtrugman) [inline src/parsers/route.cpp:34]: You're missing `clang-format on` after the comment

File: src/parsers/route.cpp [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    route route;
+
+    static const char DELIM = '\t';
```
Comment (dtrugman) [inline src/parsers/route.cpp:37]: [nitpicking] If you look at other files, the format is always:
```
... DELIM
<newline>
tokens = utils::split()
if (tokens != ...)
```

File: src/parsers/route.cpp [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    auto tokens = utils::split(line, DELIM);
+
+    if (tokens.size() != 11)
```
Comment (dtrugman) [inline src/parsers/route.cpp:40]: All the implementations use an enum to number the fields and the total count. See `net_socket.cpp` or `net_device.cpp` for examples. Please use the same here.

File: src/parsers/route.cpp [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        throw parser_error("Corrupted route - Unexpected tokens count", line);
+    }
+
+    // Parse tokens into route struct
+    route.iface          = tokens[0];
+    route.destination    = pfs::ip(pfs::ipv4(std::stoul(tokens[1], nullptr, 16)));
```
Comment (dtrugman) [inline src/parsers/route.cpp:47]: Please check out `parse_ipv4_address` in `net_socket.cpp`.

File: src/parsers/route.cpp [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    route.gateway        = pfs::ip(pfs::ipv4(std::stoul(tokens[2], nullptr, 16)));
+    route.flags          = std::stoul(tokens[3], nullptr, 16);
```
Comment (dtrugman) [inline src/parsers/route.cpp:49]: Please use `utils::stot`, it provides a cleaner API for decimals and hexadecimals and additional checks.

File: src/procfs.cpp [hunk 15]
```diff
@@ -185,4 +185,15 @@ std::string procfs::get_version_signature() const
...
 }
 
+std::vector<route> procfs::get_routes() const
+{
+    static const std::string ROUTES_FILE("net/route");
+    auto path = _root + ROUTES_FILE;
+
+    std::vector<route> output;
+    parsers::parse_lines(path, std::inserter(output, output.begin()),
+                         parsers::parse_route_line, 1);
```
Comment (dtrugman) [inline src/procfs.cpp:195]: Please use `static const size_t HEADER_LINES = 1;` just like all other implementations do.

File: src/procfs.cpp [hunk 16]
```diff
@@ -185,4 +185,15 @@ std::string procfs::get_version_signature() const
     return utils::readline(path);
 }
 
+std::vector<route> procfs::get_routes() const
+{
+    static const std::string ROUTES_FILE("net/route");
+    auto path = _root + ROUTES_FILE;
+
+    std::vector<route> output;
+    parsers::parse_lines(path, std::inserter(output, output.begin()),
```
Comment (dtrugman) [inline src/procfs.cpp:194]: You probably want `back_inserter` to avoid reversing the order of the entries.

File: src/procfs.cpp [hunk 17]
```diff
@@ -185,4 +185,15 @@ std::string procfs::get_version_signature() const
     return utils::readline(path);
 }
 
+std::vector<route> procfs::get_routes() const
+{
+    static const std::string ROUTES_FILE("net/route");
```
Comment (dtrugman) [inline src/procfs.cpp:190]: Once you move this into `net.cpp`, you won't need to specify the `net/` prefix.

File: test/route.cpp [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+TEST_CASE("Parse route", "[net][route]")
+{
+    pfs::route expected;
+
+    std::string line = "eth0\t00000000\t07030301\t0003\t0\t0\t0\t00F0FFFF\t0\t0\t0";
+
+    expected.iface       = "eth0";
+    expected.destination = pfs::ip(pfs::ipv4(0x00000000));
```
Comment (dtrugman) [inline test/route.cpp:33]: You don't need to use a double constructor, just use `pfs::ip(0x00000000)`

**Review by dtrugman — CHANGES_REQUESTED** [review]

File: include/pfs/types.hpp (hunk 3 above)
Comment (lyshar) [reply include/pfs/types.hpp:613]: I didn't see any reference to flag parsing on the library, can you give me an example from the library? not sure how a library user would like to use that

File: src/parsers/route.cpp (hunk 13 above)
Comment (lyshar) [reply src/parsers/route.cpp]: I moved `parse_ipv4_address` to `utils.cpp`, not sure if that's a good idea but it's used across multiple parsers so i thought it's a good idea, maybe create a new file `parser_utils.cpp` that includes those generic conversion functions?

File: include/pfs/types.hpp (hunk 3 above)
Comment (dtrugman) [reply include/pfs/types.hpp:613]: After discussing that offline, I think we agreed to add that there's no point in doing it now.

File: include/pfs/procfs.hpp (hunk 7 above)
Comment (dtrugman) [reply include/pfs/procfs.hpp]: The convention is to use `get_<filename>` where filename is the exact file under `/proc/net/`.
Can we use `get_route` instead of `get_routes` for consistency?

File: include/pfs/types.hpp [hunk 19]
```diff
@@ -604,20 +604,20 @@ struct id_map
...
-    unsigned short use;
-    uint32_t metric;
+    int refcnt;
+    unsigned int use;
+    int metric;
     ip mask;
-    uint32_t mtu;
-    uint32_t window;
-    uint32_t irtt;
+    unsigned int mtu;
```
Comment (dtrugman) [inline include/pfs/types.hpp:618]: `mtu` is signed, can we please change this to `int`?

**Review by dtrugman — APPROVED** [review]
